docs/tutorials/aws-sd.md

```diff
@@ -0,0 +1,200 @@
+```
+
+Verify that the namespace was truly created
+
+```console
+$ aws servicediscovery list-namespaces
+```
+
+## Deploy ExternalDNS
+
+Connect your `kubectl` client to the cluster that you want to test ExternalDNS with.
+Then apply the following manifest file to deploy ExternalDNS.
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: external-dns
+spec:
+  strategy:
+    type: Recreate
+  template:
+    metadata:
+      labels:
+        app: external-dns
+    spec:
+      containers:
+      - name: external-dns
+        image: registry.opensource.zalan.do/teapot/external-dns:v0.4.8
+        args:
+        - --source=service
+        - --source=ingress
+        - --domain-filter=external-dns-test.my-org.com # Makes ExternalDNS see only the namespaces that match the specified domain. Omit the filter if you want to process all available namespaces.
+        - --provider=aws-sd
+        - --aws-zone-type=public # Only look at public namespaces. Valid values are public, private, or no value for both)
+        - --txt-owner-id=my-identifier
+```
+
+
+## Verify that ExternalDNS works (Service example)
+
+Create the following sample application to test that ExternalDNS works.
+
+> For services ExternalDNS will look for the annotation `external-dns.alpha.kubernetes.io/hostname` on the service and use the corresponding value.
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: nginx
+  annotations:
+    external-dns.alpha.kubernetes.io/hostname: nginx.external-dns-test.my-org.com
+spec:
+  type: LoadBalancer
+  ports:
+  - port: 80
+    name: http
+    targetPort: 80
+  selector:
+    app: nginx
+
+---
+
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: nginx
+spec:
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - image: nginx
+        name: nginx
+        ports:
+        - containerPort: 80
+          name: http
+```
+
+After one minute check that a corresponding DNS record for your service was created in your hosted zone. We recommended that you use the [Amazon Route53 console](https://console.aws.amazon.com/route53) for that purpose.
+
+
+## Custom TTL
+
+The default DNS record TTL (time to live) is 300 seconds. You can customize this value by setting the annotation `external-dns.alpha.kubernetes.io/ttl`.
+For example, modify the service manifest YAML file above:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: nginx
+  annotations:
+    external-dns.alpha.kubernetes.io/hostname: nginx.external-dns-test.my-org.com
+    external-dns.alpha.kubernetes.io/ttl: 60
+spec:
+    ...
+```
+
+This will set the TTL for the DNS record to 60 seconds.
+
+
+## Clean up
+
+Delete all service objects before terminating the cluster so all load balancers get cleaned up correctly.
+
+```console
+$ kubectl delete service nginx
+```
+
+Give ExternalDNS some time to clean up the DNS records for you. Then delete the remaining service and namespace.
```

Comment:
Shouldn't the AWS SD `Service` be cleanup up by ExternalDNS when the Kubernetes `Service` is deleted? (assuming `policy=sync`)

Reply:
Good question. Theoretically it SHOULD. But I decided to simplify the provider and not provide delete service functionality. There are basically two reasons:

1. Before the service can be deleted, all instances in the given service must be deleted first. But because `DeregisterInstance` API call is asynchronous, the outcome of the operation is eventually consistent. It is needed to poll status of all `DeregisterInstance` that are in flight and only then the service can be successfully deleted. It makes things much more complicated.

2. This is reason is purely practical. User is not charged for existing services (just instances) by AWS. So there is no pressure to cleanup everything. The service can be cleaned up manually with the namespace when needed.

Honestly I am open to add this feature later if there is a demand, but I don't see any harm to release as it is now. Please let me know what do you think.

Reply:
Sounds good to me.